File: backend/shared/pricing_policy.py
from __future__ import annotations

GLOBAL_PRICING_POLICY = {
    "currency": "USD",
    "yearly_discount_pct": 20,
    "plans": {
        "free": {"id": "free", "name": "Free", "monthly": 0.0, "yearly": 0.0},
        "basic": {"id": "basic", "name": "Basic", "monthly": 5.99, "yearly": 57.50},
        "premium": {"id": "premium", "name": "Premium", "monthly": 15.99, "yearly": 153.50},
    },
}


def get_plan_amount(plan_id: str, billing_period: str = "monthly") -> float:
    normalized_plan = str(plan_id or "").strip().lower()
    normalized_period = str(billing_period or "monthly").strip().lower()
    plan = GLOBAL_PRICING_POLICY["plans"].get(normalized_plan) or GLOBAL_PRICING_POLICY["plans"]["basic"]
    if normalized_period == "yearly":
        return float(plan["yearly"])
    return float(plan["monthly"])


def get_yearly_discount_pct() -> int:
    return int(GLOBAL_PRICING_POLICY["yearly_discount_pct"])


def get_plan_name(plan_id: str) -> str:
    normalized_plan = str(plan_id or "").strip().lower()
    plan = GLOBAL_PRICING_POLICY["plans"].get(normalized_plan) or GLOBAL_PRICING_POLICY["plans"]["basic"]
    return str(plan["name"])


def get_monthly_price_label(plan_id: str) -> str:
    return f"${get_plan_amount(plan_id, 'monthly'):.2f}/mo"


def get_yearly_price_map() -> dict[str, float]:
    return {plan_id: float(plan["yearly"]) for plan_id, plan in GLOBAL_PRICING_POLICY["plans"].items()}


def get_monthly_price_map() -> dict[str, float]:
    return {plan_id: float(plan["monthly"]) for plan_id, plan in GLOBAL_PRICING_POLICY["plans"].items()}

File: frontend/src/config/pricingPolicy.ts
export const GLOBAL_PRICING_POLICY = {
  currency: 'USD',
  yearlyDiscountPct: 20,
  plans: {
    free: { id: 'free', name: 'Free', monthly: 0, yearly: 0 },
    basic: { id: 'basic', name: 'Basic', monthly: 5.99, yearly: 57.5 },
    premium: { id: 'premium', name: 'Premium', monthly: 15.99, yearly: 153.5 },
  },
} as const;

export type PricingPlanId = keyof typeof GLOBAL_PRICING_POLICY.plans;

export function getFrontendPlanPrice(planId: string, billingPeriod: 'monthly' | 'yearly' = 'monthly') {
  const normalized = String(planId || '').toLowerCase() as PricingPlanId;
  const plan = GLOBAL_PRICING_POLICY.plans[normalized] || GLOBAL_PRICING_POLICY.plans.basic;
  return billingPeriod === 'yearly' ? plan.yearly : plan.monthly;
}

export function getFrontendPlanName(planId: string) {
  const normalized = String(planId || '').toLowerCase() as PricingPlanId;
  return (GLOBAL_PRICING_POLICY.plans[normalized] || GLOBAL_PRICING_POLICY.plans.basic).name;
}

export function getFrontendYearlyDiscountPct() {
  return GLOBAL_PRICING_POLICY.yearlyDiscountPct;
}

export function buildFrontendPricingSaveLabel(template: string | undefined, fallback: string = 'SAVE {pct}%') {
  const safeTemplate = String(template || fallback || '').trim() || fallback;
  return safeTemplate.replace('{pct}', String(getFrontendYearlyDiscountPct()));
}